From 930e35c5f5ff24e01de295bfee3c73d4fe591282 Mon Sep 17 00:00:00 2001
From: wengjun1 <wengjun1@motorola.com>
Date: Thu, 10 Dec 2015 18:27:17 +0800
Subject: [PATCH 549/959] IKSWM-13820 input: synaptics_dsx_i2c: add in progress
 blank notification

Some display/touch panels have unique requirements for sleep sequence.
Notification FB_IN_PROGRESS_EVENT_BLANK is sent in the middle of
display power off sequence and shall be used instead of FB_EARLY_EVENT_BLANK
for such panels.

Change-Id: I594b12175e37db114cacce40006d4ff6330aeb31
Signed-off-by: wengjun1 <wengjun1@motorola.com>
Reviewed-on: http://gerrit.mot.com/806279
SLTApproved: Slta Waiver <sltawvr@motorola.com>
SME-Granted: SME Approvals Granted
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Konstantin Makariev <kmakariev@motorola.com>
Reviewed-by: Igor Kovalenko <igork@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/input/touchscreen/synaptics_dsx_i2c.c | 13 ++++++++++---
 drivers/input/touchscreen/synaptics_dsx_i2c.h |  1 +
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/drivers/input/touchscreen/synaptics_dsx_i2c.c b/drivers/input/touchscreen/synaptics_dsx_i2c.c
index c512a9e25b2..49b1f128774 100644
--- a/drivers/input/touchscreen/synaptics_dsx_i2c.c
+++ b/drivers/input/touchscreen/synaptics_dsx_i2c.c
@@ -1480,6 +1480,11 @@ static struct synaptics_dsx_platform_data *
 	} else
 		dev_err(&client->dev, "clip area read failure\n");
 
+	if (of_property_read_bool(np, "synaptics,use-in-progress-event-blank")) {
+		pr_notice("using in progress event blank\n");
+		rmi4_data->event_blank = FB_IN_PROGRESS_EVENT_BLANK;
+	}
+
 	return pdata;
 }
 #else
@@ -6398,6 +6403,8 @@ static int synaptics_rmi4_probe(struct i2c_client *client,
 		return retval;
 	}
 
+	rmi4_data->event_blank = FB_EARLY_EVENT_BLANK;
+
 	if (client->dev.of_node)
 		platform_data = synaptics_dsx_of_init(client, rmi4_data);
 	else
@@ -6793,15 +6800,15 @@ static int synaptics_dsx_panel_cb(struct notifier_block *nb,
 	struct synaptics_rmi4_data *rmi4_data =
 		container_of(nb, struct synaptics_rmi4_data, panel_nb);
 
-	if ((event == FB_EARLY_EVENT_BLANK || event == FB_EVENT_BLANK) &&
+	if ((event == rmi4_data->event_blank || event == FB_EVENT_BLANK) &&
 			evdata && evdata->info && evdata->info->node == 0 &&
 			evdata->data && rmi4_data) {
 		int *blank = evdata->data;
 		pr_debug("fb notification: event = %lu blank = %d\n", event, *blank);
-		/* entering suspend upon early blank event */
+		/* entering suspend upon early blank or in-progress blank event*/
 		/* to ensure shared power supply is still on */
 		/* for in-cell design touch solutions */
-		if (event == FB_EARLY_EVENT_BLANK) {
+		if (event == rmi4_data->event_blank) {
 			if (*blank != FB_BLANK_POWERDOWN)
 				return 0;
 			synaptics_dsx_display_off(&rmi4_data->i2c_client->dev);
diff --git a/drivers/input/touchscreen/synaptics_dsx_i2c.h b/drivers/input/touchscreen/synaptics_dsx_i2c.h
index bcaa1c1707c..37a7e8c26e3 100644
--- a/drivers/input/touchscreen/synaptics_dsx_i2c.h
+++ b/drivers/input/touchscreen/synaptics_dsx_i2c.h
@@ -550,6 +550,7 @@ struct synaptics_rmi4_data {
 	bool clipping_on;
 	struct synaptics_clip_area *clipa;
 	struct mutex rmi4_exp_init_mutex;
+	unsigned long event_blank;
 };
 
 struct time_keeping {
-- 
2.11.0

